content/guides/java/inject-extract.md

```diff
@@ -2,8 +2,142 @@
+        if (parentSpan == null) {
+            spanBuilder = tracer.buildSpan(operationName);
+        } else {
+            spanBuilder = tracer.buildSpan(operationName).asChildOf(parentSpan);
+        }
+    } catch (IllegalArgumentException e) {
+        spanBuilder = tracer.buildSpan(operationName);
+    }
+    return spanBuilder.withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER).startActive(true);
+}
+```
+
+In the above example, instead of using a dedicated adapter class to convert JAXRS `HttpHeaders` type into `io.opentracing.propagation.TextMap`, the headers are copied to a plain `HashMap<String, String>` and converted using a standard adapter `TextMapExtractAdapter`.
+
+## Injecting and Extracting TextMap
+The process of injecting and extracting `TextMap` is similar to that of HTTP. Given below are some elaborated code examples of injecting and extracting tracing information using `TextMap` format.
+
+```
+protected void attachTraceInfo(Tracer tracer, Span span, final Invocation inv) {
+    tracer.inject(span.context(), Format.Builtin.TEXT_MAP, new TextMap() {
+
+        @Override
+        public void put(String key, String value) {
+            inv.getAttachments().put(key, value);
+        }
+
+        @Override
+        public Iterator<Map.Entry<String, String>> iterator() {
+            throw new UnsupportedOperationException("TextMapInjectAdapter should only be used with Tracer.inject()");
+        }
+    });
+}
+```
+The above method injects the invocation information into the span’s context on the consumer side. The `TextMap` interface has been implemented to create a `TextMap` of invocation attachments.
+
+```
+protected Span extractTraceInfo(Tracer tracer, Invoker<?> invoker, Invocation inv) {
+
+    Tracer.SpanBuilder span = tracer.buildSpan(“Operation_Name_Here”);
+    try {
+        SpanContext spanContext = tracer.extract(Format.Builtin.TEXT_MAP, new TextMapExtractAdapter(inv.getAttachments()));
+        if (spanContext != null) {
+            //if spanContext is extracted, the spanContext is propagated to the new span
+              
+            span.asChildOf(spanContext);
+        }
+    } catch (Exception e) {
+        span.withTag("Error", "extract from request fail, error msg:" + e.getMessage());
+    }
+    return span.startManual();
+}
+```
+On the provider side, `tracer.extract()` is used to extract the invocation attachments. The logic is similar to `tracer.inject()`.
+
+
+
+## Injecting and Extracting Binary
+Binary format is used for injecting/extracting span context encoded in an opaque byte array. Opaque means that the inner representation of the array is not known.
+
+Injecting and extracting binary can be implemented as follows:
+
+```java
+//Inject binary
+public void byteBufferInjection() throws Exception {
+  byte[] key = "foo".getBytes(ByteBufferContext.CHARSET), value = "bar".getBytes(ByteBufferContext.CHARSET);
+  ByteBuffer byteBuffer = ByteBuffer.allocate(2 + 2 * 4 + key.length + value.length);
```

Comment:
Same as above: add a link to a working example. I myself would not know what those magic numbers are...